File: pset6/vigenere.py
import sys

def printCipher(text, key):
    encryptedString = '';
    k = 0;
    # for each character check the of upper or lowercase
    for i in range(len(text)):
        # only encrypt letters, other characters remain unchanged
        if text[i].isalpha():
            if text[i].isupper():
                startIndex = ord('A')
            else:
                startIndex = ord('a')
            # cipher is (text[i] - k) mod 26
            encryptedString += chr(startIndex + (ord(text[i]) - startIndex + ord( key[k % len(key)].lower()) - ord('a')) % 26)
        else:
            encryptedString += text[i]
        k += 1
    return encryptedString
        
def main():
    # check if number of arguments is correct
    if len(sys.argv) != 2:
        print("Usage: python vigenere.py k")
        exit(1)
    
    # check if the key only contains characters
    key = sys.argv[1]
    if key.isalpha() != True:
        print("Usage: python vigenere.py k")
        exit(1)
    
    # request the plaintext
    plain = input("plaintext: ")
    
    # print the cipher
    print("ciphertext: ", printCipher(plain, key))
    exit(0)

if __name__ == "__main__":
    main()

File: pset2/vigenere.c
#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <string.h>

void printCipher(string text, string key)
{
    // loop over each character
    for(int i = 0, k = 0, n = strlen(text); i < n; i++)
    {
        // check if the character is alphanumerc
        if(isalpha(text[i]))
        {
            // determine start index (upper or lower case)
            int startIndex;
            if(isupper(text[i]))
            {
                startIndex = 'A';
            }
            else
            {
                startIndex = 'a';
            }
            // print the cipher: startIndex + (char - startIndex + key) % 26,
            // where key corresponds to the nth alphanumeric letter in the plaintext modulus the keylength 
            printf("%c", startIndex + (text[i] - startIndex + tolower(key[k % strlen(key)]) - 'a') % 26);
            k++;
        }
        else
        {
            // non-alphanumeric, just return the character
            printf("%c", text[i]);
        }
    }
    printf("\n");
}

bool checkAlpha(string s)
{
    // loop over each character in string s to check if the string contains any non-alphanumeric characters
    for(int i = 0, n = strlen(s); i < n; i++){
        if(!isalpha(s[i]))
            return false;
    }
    return true;
}

int main(int argc, string argv[])
{
    // check if the key (k) is provided
    if(argc != 2)
    {
        printf("Usage: ./vigenere k\n");
        return 1;
    }
    
    // check if the key (k) only contains alphanumeric characters
    if(!checkAlpha(argv[1]))
    {
        printf("Usage: ./vigere k\n");
        return 1;
    }
    
    // ask for plaintext to convert
    printf("plaintext: ");
    string plain = get_string();
    
    //print ciphertext
    printf("ciphertext: ");
    printCipher(plain, argv[1]);
    
    return 0;
}